Add assume_safety_prior_to_version config option

vergenzt opened the pull request:
Fixes #12 and #3.

Let me know what you think!

vergenzt commented:
Adding changelog and readme entries right now, will amend.

ankane commented:
Hey @vergenzt, thanks for the PR 👍  It looks like it's causing the tests for fail. Also, left a few comments inline.

ankane commented on test/strong_migrations_test.rb:
```diff
@@ -207,6 +221,18 @@ def test_create_table_force
+  def test_new_remove_column_not_assumed_safe
+    StrongMigrations.assume_safety_prior_to_version = SAFETY_ASSUMED_VERSION_CUTOFF
+    assert_unsafe NewRemoveColumnNotAssumedSafe
+    StrongMigrations.assume_safety_prior_to_version = 0
```
Let's use `ensure` here.

ankane commented on CHANGELOG.md:
```diff
@@ -32,3 +33,5 @@
 
 - First release
+
+[@vergenzt]: https://github.com/vergenzt
```
The current changelog doesn't list contributors. Let's keep it consistent.

vergenzt replied:
Sorry about that. Last repo I contributed to did this and I didn't think about it.

vergenzt commented:
Thanks for the review @ankane! As I was looking at the tests I realized that success or failure of the tests depended on execution order, so I refactored the tests a bit in 2b005e6 to wipe the database and create the users table before each test case so nothing bleeds through. I believe I've addressed all comments.

Lmk what you think!

ankane commented:
Thanks for addressing. There should be a way to make tests work without wiping the DB (as they do now). I'd prefer to keep the changeset to a minimum.

vvondra commented:
This would be much appreciated!

ankane commented:
Hey @vergenzt, I cleaned this up and merged in 50f9827d5bc580dae5ec8c4c24bd822f283e0ee3. Thanks for driving this. I think it'll be a very useful feature. 👍

nisanth074 commented on lib/strong_migrations/migration.rb:
```diff
@@ -69,6 +69,14 @@ def method_missing(method, *args, &block)
+    end
+
+    def should_assume_safety_due_to_version?
+      version && version <= StrongMigrations.assume_safety_prior_to_version
```
@vergenzt @ankane In what cases is `version` nil?

I just added the strong_migrations gem to our Rails 3 app. Unfortunately, the `start_after` option doesn't seem to work. Old dangerous migrations fail despite me configuring `start_after` option https://gist.github.com/nisanth074/ea45c8cdd50d4a08b53cb97b91fc3a6b

I added a `binding.pry` statement above this line and ran `bundle exec rake db:create db:migrate` again. It seems like `version` is always `nil` for all our migrations. So, `should_assume_safety_due_to_version?` returns nil and strong_migrations raises an error.

Can you let me know in what cases is `version` nil? Is Rails 3 supported?

Thanks for your time.

ijones16 commented:
Thanks for your work on this feature. It helped me out with a 5 year old project 🙌

jpwhitaker commented:
I just found this thread after installing [rpush](https://github.com/rpush/rpush).

I had 28 migrations that pass the strong_migration check, and the rpush installation added 17 more that fail.  Is there any method like this that would allow me to skip a range of migrations?

![Screen Shot 2021-01-18 at 11 45 02 AM](https://user-images.githubusercontent.com/2548116/104965331-9c6de380-5982-11eb-859f-a171ade98e43.png)
